### PR TITLE
chore(netlify): use Node.js for testing

### PR DESCRIPTION
**Pull request by abidjappie**
## Changes

- This PR migrates the `@astro/netlify` tests to Node.js (original issue: https://github.com/withastro/adapters/issues/144)
- Will include changeset once it's completed `pnpm exec changeset`

### Challenges
- Node.js does not support snapshot matching for now (https://github.com/nodejs/node/issues/48260)
- Currently investigating failing test related to `fs`

## Testing

For now, run the following command in `packages/netlify`
```zsh
node --test **/*.test.js
```

## Docs

For now there is no documentation, but that might change

### CODE REVIEW

**Comment by alexanderniebuhr** [comment]
@abidjappie Thanks for the PR. Maybe you can consult with @Skn0tt to see if we can rewrite the snapshot test to be still valuable without a snapshot :)

**Comment by abidjappie** [comment]
Something I noticed is that either `Image CDN` or `Middleware` tests will fail when running on local (when one passes the other fails). I suspect it's related to how it builds the fixtures... in either case, it cannot find the file.

`Image CDN`:
```
[Error: ENOENT: no such file or directory, open '/Users/abid.jappie/code/withastro-adapters/packages/netlify/test/functions/fixtures/middleware/dist/_astro/astronaut.jcSXTCiO_1Hy9L1.webp'] { errno: -2, code: 'ENOENT', syscall: 'open', path: '/Users/abid.jappie/code/withastro-adapters/packages/netlify/test/functions/fixtures/middleware/dist/_astro/astronaut.jcSXTCiO_1Hy9L1.webp' }
```

`Middleware`:
```
 [Error: ENOENT: no such file or directory, open '/Users/abid.jappie/code/withastro-adapters/packages/netlify/test/functions/fixtures/middleware/dist/_astro/astronaut.jcSXTCiO_1Hy9L1.webp'] { errno: -2, code: 'ENOENT', syscall: 'open', path: '/Users/abid.jappie/code/withastro-adapters/packages/netlify/test/functions/fixtures/middleware/dist/_astro/astronaut.jcSXTCiO_1Hy9L1.webp' }
```

File: packages/netlify/test/functions/redirects.test.js [hunk 1]
```diff
@@ -13,8 +14,10 @@ describe('SSR - Redirects', () => {
 	it('Creates a redirects file', async () => {
 		const redirects = await fixture.readFile('./_redirects');
 		const parts = redirects.split(/\s+/);
-		expect(parts).to.deep.equal(['', '/other', '/', '301', '']);
-		expect(redirects).to.matchSnapshot();
+		assert.deepEqual(parts,['', '/other', '/', '301', '']);
+		// TODO: Snapshots are not supported in Node.js test yet (https://github.com/nodejs/node/issues/48260)
+		// expect(redirects).to.matchSnapshot();
```
Comment (Skn0tt) [inline packages/netlify/test/functions/redirects.test.js:19]: could we inline the snapshot, and do `assert.equal(redirects, "....")`?

**Review by Skn0tt — COMMENTED** [review]
looks good on first sight!

**Comment by Skn0tt** [comment]
> it cannot find the file.

Hmm, I wonder wether that also happens on CI 🤔 Mind sharing the full stacktraces, to see where the `fs.open` call originates?

**Comment by alexanderniebuhr** [comment]
I'll also migrate this script https://github.com/withastro/astro/blob/main/scripts/cmd/test.js over to the adapters repo, because using the Node's cli has some downsides. Once I've migrated it, we can use it for this PR too.

**Comment by alexanderniebuhr** [comment]
@abidjappie, @Skn0tt  What is the state? Do we need still to figure stuff out, or can we update this PR with latest main and get this merged?

I also just merged https://github.com/withastro/adapters/pull/150 which adds support for `astro-script test`

File: packages/netlify/test/functions/redirects.test.js (hunk 1 above)
Comment (abidjappie) [reply packages/netlify/test/functions/redirects.test.js:19]: I'll give it a bash. 🙇 thank you for the feedback!

In that case we won't continue generating snapshots, which should be okay if this pages content doesn't change often?

**Comment by abidjappie** [comment]
> > it cannot find the file.
> 
> Hmm, I wonder wether that also happens on CI 🤔 Mind sharing the full stacktraces, to see where the `fs.open` call originates?

```zsh
✖ enables Netlify Image CDN (295.416042ms)
  Error: ENOENT: no such file or directory, stat '/Users/abid.jappie/code/withastro-adapters/packages/netlify/test/functions/fixtures/middleware/.netlify/functions-internal/ssr/entry.mjs'
      at Object.statSync (node:fs:1690:3)
      at removeEmptyDirs (file:///Users/abid.jappie/code/withastro-adapters/node_modules/.pnpm/astro@4.2.0_@types+node@18.19.2_typescript@5.3.2/node_modules/astro/dist/core/fs/index.js:12:11)
      at file:///Users/abid.jappie/code/withastro-adapters/node_modules/.pnpm/astro@4.2.0_@types+node@18.19.2_typescript@5.3.2/node_modules/astro/dist/core/fs/index.js:18:7
      at Array.map (<anonymous>)
      at removeEmptyDirs (file:///Users/abid.jappie/code/withastro-adapters/node_modules/.pnpm/astro@4.2.0_@types+node@18.19.2_typescript@5.3.2/node_modules/astro/dist/core/fs/index.js:16:11)
      at cleanStaticOutput (file:///Users/abid.jappie/code/withastro-adapters/node_modules/.pnpm/astro@4.2.0_@types+node@18.19.2_typescript@5.3.2/node_modules/astro/dist/core/build/static-build.js:314:5)
      at async staticBuild (file:///Users/abid.jappie/code/withastro-adapters/node_modules/.pnpm/astro@4.2.0_@types+node@18.19.2_typescript@5.3.2/node_modules/astro/dist/core/build/static-build.js:104:7)
      at async AstroBuilder.build (file:///Users/abid.jappie/code/withastro-adapters/node_modules/.pnpm/astro@4.2.0_@types+node@18.19.2_typescript@5.3.2/node_modules/astro/dist/core/build/index.js:138:5)
      at async AstroBuilder.run (file:///Users/abid.jappie/code/withastro-adapters/node_modules/.pnpm/astro@4.2.0_@types+node@18.19.2_typescript@5.3.2/node_modules/astro/dist/core/build/index.js:169:7)
      at async build (file:///Users/abid.jappie/code/withastro-adapters/node_modules/.pnpm/astro@4.2.0_@types+node@18.19.2_typescript@5.3.2/node_modules/astro/dist/core/build/index.js:47:3) {
    errno: -2,
    syscall: 'stat',
    code: 'ENOENT',
    path: '/Users/abid.jappie/code/withastro-adapters/packages/netlify/test/functions/fixtures/middleware/.netlify/functions-internal/ssr/entry.mjs'
  }
```

**Comment by abidjappie** [comment]
@alexanderniebuhr I'll look into the feedback sometime today, and try to get some stack traces.
After that I would like to run the CI tests so we can see if it errors out or not before proceeding 🙇

**Comment by abidjappie** [comment]
@Skn0tt 
> Hmm, I wonder wether that also happens on CI 🤔 Mind sharing the full stacktraces, to see where the `fs.open` call originates?

I'm running the tests via `node --test`. In my attempt to find more information I surrounded both the test case, and the before each with a catch... and it seems to be occurring in the `beforeEach` level. Specifically:
```ts
await fixture.build();
```

My guess is that the test are running concurrently, which is why it fails in different places each time. I wrapped it in `console.trace`:
```
Trace: Error: ENOENT: no such file or directory, stat '/Users/abid.jappie/code/withastro-adapters/packages/netlify/test/functions/fixtures/middleware/.netlify/functions-internal/ssr/manifest_n9sb-Uc7.mjs'
    at Object.statSync (node:fs:1690:3)
    at removeEmptyDirs (file:///Users/abid.jappie/code/withastro-adapters/node_modules/.pnpm/astro@4.3.5_@types+node@18.19.2_typescript@5.3.2/node_modules/astro/dist/core/fs/index.js:12:11)
    at file:///Users/abid.jappie/code/withastro-adapters/node_modules/.pnpm/astro@4.3.5_@types+node@18.19.2_typescript@5.3.2/node_modules/astro/dist/core/fs/index.js:18:7
    at Array.map (<anonymous>)
    at removeEmptyDirs (file:///Users/abid.jappie/code/withastro-adapters/node_modules/.pnpm/astro@4.3.5_@types+node@18.19.2_typescript@5.3.2/node_modules/astro/dist/core/fs/index.js:16:11)
    at cleanStaticOutput (file:///Users/abid.jappie/code/withastro-adapters/node_modules/.pnpm/astro@4.3.5_@types+node@18.19.2_typescript@5.3.2/node_modules/astro/dist/core/build/static-build.js:314:5)
    at async staticBuild (file:///Users/abid.jappie/code/withastro-adapters/node_modules/.pnpm/astro@4.3.5_@types+node@18.19.2_typescript@5.3.2/node_modules/astro/dist/core/build/static-build.js:104:7)
    at async AstroBuilder.build (file:///Users/abid.jappie/code/withastro-adapters/node_modules/.pnpm/astro@4.3.5_@types+node@18.19.2_typescript@5.3.2/node_modules/astro/dist/core/build/index.js:138:5)
    at async AstroBuilder.run (file:///Users/abid.jappie/code/withastro-adapters/node_modules/.pnpm/astro@4.3.5_@types+node@18.19.2_typescript@5.3.2/node_modules/astro/dist/core/build/index.js:169:7)
    at async build (file:///Users/abid.jappie/code/withastro-adapters/node_modules/.pnpm/astro@4.3.5_@types+node@18.19.2_typescript@5.3.2/node_modules/astro/dist/core/build/index.js:47:3) {
  errno: -2,
  syscall: 'stat',
  code: 'ENOENT',
  path: '/Users/abid.jappie/code/withastro-adapters/packages/netlify/test/functions/fixtures/middleware/.netlify/functions-internal/ssr/manifest_n9sb-Uc7.mjs'
}
    at TestContext.<anonymous> (file:///Users/abid.jappie/code/withastro-adapters/packages/netlify/test/functions/edge-middleware.test.js:18:13)
    at async TestHook.run (node:internal/test_runner/test:574:9)
    at async node:internal/test_runner/test:509:9
    at async Suite.runHook (node:internal/test_runner/test:507:7)
    at async Test.run (node:internal/test_runner/test:549:9)
    at async Suite.processPendingSubtests (node:internal/test_runner/test:318:7)
```

**Comment by Skn0tt** [comment]
Could be because of concurrency, yes! I see you replaced `before` with `beforeEach` in the edge-middleware test, what happens when you switch that back to `before`?

**Comment by abidjappie** [comment]
@alexanderniebuhr can you help me to get the test running using `astro-scripts`. 
I pulled latest branch and tried moving to it, but I get no output:

```zsh
> @astrojs/netlify@5.1.1 test /Users/abid.jappie/code/withastro-adapters/packages/netlify
> pnpm run test-fn && pnpm run test-static

> @astrojs/netlify@5.1.1 test-fn /Users/abid.jappie/code/withastro-adapters/packages/netlify
> astro-scripts --test ./test/functions/*.test.js

> @astrojs/netlify@5.1.1 test-static /Users/abid.jappie/code/withastro-adapters/packages/netlify
> astro-scripts --test ./test/static/*.test.js
```

`node --test **/*.test.js` works fine.

**Comment by abidjappie** [comment]
@Skn0tt 
I managed to get `edge-middleware.test.js` to pass, there are two parts:
1. You are totally right that changing `before` to `beforeEach` was part of the solution.
2. We also need to run the `await fixture.clean()` in the `after`

Unfortunately, I think this might solve the issue of file access overlapping in a single test, I think that different test files can still have this effect on one another: `edge-middleware` now consistently passes, but `image-cdn` consistently fails.

I can easily break `edge-middleware` by `fixture.clean` inside of `image-cdn.test.js`.
Is there someway to add a test specific config to `test-utils.js` or can we somehow isolate the tests/ ensure they don't run in parallel?

This also works for running `image-cdn.test.js` in isolation: `await fixture.clean()` after each `it`:
```zsh
node --test **/image-cdn.test.js
▶ Image CDN
  ▶ when running outside of netlify
    ✔ does not enable Image CDN (714.118083ms)
  ▶ when running outside of netlify (714.634083ms)

  ▶ when running inside of netlify
    ✔ enables Netlify Image CDN (326.024042ms)
    ✔ respects image CDN opt-out (434.420417ms)
  ▶ when running inside of netlify (760.604291ms)

▶ Image CDN (1475.79075ms)

ℹ tests 3
ℹ suites 3
ℹ pass 3
ℹ fail 0
ℹ cancelled 0
ℹ skipped 0
ℹ todo 0
ℹ duration_ms 2056.534583
```

**Comment by alexanderniebuhr** [comment]
@abidjappie let me check.

File: packages/netlify/package.json [hunk 2]
```diff
@@ -30,10 +30,10 @@
   ],
   "scripts": {
     "build": "tsc",
-    "test-fn": "mocha --exit --timeout 20000 --file \"./test/setup.js\" test/functions/",
-    "test-static": "mocha --exit --timeout 20000 --file \"./test/setup.js\" test/static/",
-    "test": "pnpm test-fn && pnpm test-static",
-    "test:hosted": "mocha --exit --timeout 30000 test/hosted"
+    "test-fn": "astro-scripts --test ./test/functions/*.test.js",
```
Comment (alexanderniebuhr) [inline packages/netlify/package.json:33]:
```suggestion
    "test-fn": "astro-scripts test ./test/functions/*.test.js",
```

File: packages/netlify/package.json [hunk 3]
```diff
@@ -30,10 +30,10 @@
...
+    "test-static": "astro-scripts --test ./test/static/*.test.js",
```
Comment (alexanderniebuhr) [inline packages/netlify/package.json:34]:
```suggestion
    "test-static": "astro-scripts test ./test/static/*.test.js",
```

File: packages/netlify/package.json [hunk 4]
```diff
@@ -30,10 +30,10 @@
...
+    "test": "pnpm run test-fn && pnpm run test-static",
+    "test:hosted": "astro-scripts --test ./test/hosted/*.test.js"
```
Comment (alexanderniebuhr) [inline packages/netlify/package.json:36]:
```suggestion
    "test:hosted": "astro-scripts test ./test/hosted/*.test.js"
```

**Comment by abidjappie** [comment]
@Skn0tt 
> Hmm, I'm not sure I follow. Could you push those changes? then I can try running it locally to see what's going wrong.

You can try with this branch. State: `edge-middleware` should pass, `image-cdn` seems to fail.
https://github.com/withastro/adapters/pull/158/commits/f26386e93078232cf992f934a78908549b10c537

If you surround each case in `image-cdn` with `test.only` you can of course get it to pass.

File: packages/netlify/package.json (hunk 4 above)
Comment (abidjappie) [reply packages/netlify/package.json]: Thank you 🙇 I added it to https://github.com/withastro/adapters/pull/158/commits/f26386e93078232cf992f934a78908549b10c537 but I see you also pushed it 😂

**Comment by alexanderniebuhr** [comment]
This is interesting, using `astro-scripts test` does not run tests in parallel. See: https://github.com/withastro/adapters/blob/main/scripts/cmd/test.js#L42-L54

**Comment by abidjappie** [comment]
@alexanderniebuhr true. I also noticed that running with `astro-scripts test ./test/functions/*.test.js` only runs the first test in the folder.
```zsh
pnpm run test-fn

> @astrojs/netlify@5.1.2 test-fn /Users/abid.jappie/code/withastro-adapters/packages/netlify
> astro-scripts test ./test/functions/*.test.js

▶ Cookies
  ✔ Can set multiple (13.354375ms)
  ✔ renders dynamic 404 page (4.59525ms)
▶ Cookies (597.759083ms)

ℹ tests 2
ℹ suites 1
ℹ pass 2
ℹ fail 0
ℹ cancelled 0
ℹ skipped 0
ℹ todo 0
ℹ duration_ms 1190.360458
```

I'm not sure if it supports recursive, but this works for me:
```
    "test-fn": "for f in test/functions/*.test.js; do echo $f; astro-scripts test $f; done"
```

**Comment by abidjappie** [comment]
@Skn0tt I've pushed some changes now. You can get all tests to pass.
```zsh
for f in test/functions/*.test.js; do echo $f; pnpm astro-scripts test $f; done
```

It may be related to https://nodejs.org/api/test.html#extraneous-asynchronous-activity since removing the `env` cleanup in https://github.com/withastro/adapters/pull/158/commits/afa37eaa7bb45de556d677778e8f3bc2cb611c5d somehow fixed it.

I think it may need further investigation at some point, but I'm just relieved it's behaving somewhat consistently on my machine now.

File: packages/netlify/test/functions/redirects.test.js [hunk 5]
```diff
@@ -13,8 +14,9 @@ describe('SSR - Redirects', () => {
 	it('Creates a redirects file', async () => {
 		const redirects = await fixture.readFile('./_redirects');
 		const parts = redirects.split(/\s+/);
-		expect(parts).to.deep.equal(['', '/other', '/', '301', '']);
-		expect(redirects).to.matchSnapshot();
+		assert.deepEqual(parts,['', '/other', '/', '301', '']);
+		// Snapshots are not supported in Node.js test yet (https://github.com/nodejs/node/issues/48260)
+		assert.equal(redirects,'\n/other    /       301\n');
```
Comment (Skn0tt) [inline packages/netlify/test/functions/redirects.test.js:19]: There might be a file left on disk that contains the (now unused) snapshot contents. Should we remove that?

**Review by Skn0tt — APPROVED** [review]
Looks wonderful! Good job at getting all the tests to pass under a new test runner :D

Left a tiny nit, but nothing blocking.

File: packages/netlify/test/functions/redirects.test.js (hunk 5 above)
Comment (alexanderniebuhr) [reply packages/netlify/test/functions/redirects.test.js:19]: If there are any files which we don't need anymore, we should remove it!

File: packages/netlify/package.json [hunk 6]
```diff
@@ -30,10 +30,10 @@
   ],
   "scripts": {
     "build": "tsc",
-    "test-fn": "mocha --exit --timeout 20000 --file \"./test/setup.js\" test/functions/",
-    "test-static": "mocha --exit --timeout 20000 --file \"./test/setup.js\" test/static/",
-    "test": "pnpm test-fn && pnpm test-static",
-    "test:hosted": "mocha --exit --timeout 30000 test/hosted"
+    "test-fn": "astro-scripts test ./test/functions/*.test.js",
+    "test-static": "astro-scripts test ./test/static/*.test.js",
```
Comment (alexanderniebuhr) [inline packages/netlify/package.json:34]:
```suggestion
    "test-static": "astro-scripts test \"test/static/*.test.js\"",
```

File: packages/netlify/package.json [hunk 7]
```diff
@@ -30,10 +30,10 @@
...
+    "test": "pnpm run test-fn && pnpm run test-static",
+    "test:hosted": "astro-scripts test ./test/hosted/*.test.js"
```
Comment (alexanderniebuhr) [inline packages/netlify/package.json:36]:
```suggestion
    "test:hosted": "astro-scripts test \"test/hosted/*.test.js\""
```

**Comment by alexanderniebuhr** [comment]
@abidjappie Thank you so much for your contribution. Appreciate the help ❤️

**Review by ematipico — APPROVED** [review]

**Pull request merged**